File: apps/api/app/services/todos/recurrence.py
"""Simple Schedule repeats — daily / weekdays / weekly / monthly."""

from __future__ import annotations

from calendar import monthrange
from datetime import UTC, datetime, timedelta
from typing import Literal, TypeGuard
from zoneinfo import ZoneInfo

RecurrenceRule = Literal["daily", "weekdays", "weekly", "monthly"]
RECURRENCE_RULES: tuple[RecurrenceRule, ...] = (
    "daily",
    "weekdays",
    "weekly",
    "monthly",
)


def is_recurrence_rule(value: object) -> TypeGuard[RecurrenceRule]:
    return value in RECURRENCE_RULES


def _zone(tz_name: str | None) -> ZoneInfo:
    try:
        return ZoneInfo(tz_name or "UTC")
    except Exception:
        return ZoneInfo("UTC")


def _as_local(dt: datetime, tz_name: str | None) -> datetime:
    tz = _zone(tz_name)
    if dt.tzinfo is None:
        return dt.replace(tzinfo=tz)
    return dt.astimezone(tz)


def _add_month(local: datetime) -> datetime:
    month = local.month + 1
    year = local.year + (month - 1) // 12
    month = (month - 1) % 12 + 1
    day = min(local.day, monthrange(year, month)[1])
    return local.replace(year=year, month=month, day=day)


def _step_local(local: datetime, rule: RecurrenceRule) -> datetime:
    if rule == "daily":
        return local + timedelta(days=1)
    if rule == "weekly":
        return local + timedelta(days=7)
    if rule == "monthly":
        return _add_month(local)
    nxt = local + timedelta(days=1)
    while nxt.weekday() >= 5:
        nxt += timedelta(days=1)
    return nxt


def snap_first_due(
    due_at: datetime, rule: RecurrenceRule | None, *, timezone: str | None
) -> datetime:
    """Weekday repeats skip Sat/Sun for the first fire."""
    if rule != "weekdays":
        return due_at
    local = _as_local(due_at, timezone)
    while local.weekday() >= 5:
        local += timedelta(days=1)
    return local.astimezone(due_at.tzinfo or _zone(timezone))


def next_recurring_due(
    due_at: datetime,
    rule: RecurrenceRule,
    *,
    now: datetime,
    timezone: str | None,
) -> datetime:
    """Advance *due_at* until it is strictly after *now* in the user's timezone."""
    local = _as_local(due_at, timezone)
    now_local = _as_local(now, timezone)
    if local.astimezone(UTC) > now_local.astimezone(UTC):
        return due_at
    elapsed_days = max(0, (now_local.date() - local.date()).days)
    if rule == "monthly":
        months = max(0, (now_local.year - local.year) * 12 + now_local.month - local.month)
        # Preserve existing sequential month-end clamping, including skipped
        # Februarys, without walking every missed occurrence. Every 48-month
        # interval includes a non-leap February (the shortest possible month).
        day = local.day
        for offset in range(1, min(months, 48) + 1):
            index = local.year * 12 + local.month - 1 + offset
            day = min(day, monthrange(index // 12, index % 12 + 1)[1])
        index = local.year * 12 + local.month - 1 + months
        nxt = local.replace(year=index // 12, month=index % 12 + 1, day=day)
    else:
        days = elapsed_days // 7 * 7 if rule == "weekly" else elapsed_days
        nxt = local + timedelta(days=days)
        if rule == "weekdays":
            while nxt.weekday() >= 5:
                nxt += timedelta(days=1)
    if nxt.astimezone(UTC) <= now_local.astimezone(UTC):
        nxt = _step_local(nxt, rule)
    return nxt.astimezone(due_at.tzinfo or _zone(timezone))

File: apps/mobile/lib/todos/recurrence.ts
export const RECURRENCE_RULES = ["daily", "weekdays", "weekly", "monthly"] as const;
export type RecurrenceRule = (typeof RECURRENCE_RULES)[number];

export function isRecurrenceRule(value: unknown): value is RecurrenceRule {
  return (
    typeof value === "string" &&
    (RECURRENCE_RULES as readonly string[]).includes(value)
  );
}

function addMonth(due: Date): Date {
  const next = new Date(due);
  const day = next.getDate();
  next.setMonth(next.getMonth() + 1);
  if (next.getDate() < day) {
    next.setDate(0);
  }
  return next;
}

function stepLocal(due: Date, rule: RecurrenceRule): Date {
  const next = new Date(due);
  if (rule === "daily") {
    next.setDate(next.getDate() + 1);
    return next;
  }
  if (rule === "weekly") {
    next.setDate(next.getDate() + 7);
    return next;
  }
  if (rule === "monthly") {
    return addMonth(due);
  }
  do {
    next.setDate(next.getDate() + 1);
  } while (next.getDay() === 0 || next.getDay() === 6);
  return next;
}

export function snapFirstDue(due: Date, rule: RecurrenceRule | null): Date {
  if (rule !== "weekdays") return due;
  const next = new Date(due);
  while (next.getDay() === 0 || next.getDay() === 6) {
    next.setDate(next.getDate() + 1);
  }
  return next;
}

export function nextRecurringDue(
  due: Date,
  rule: RecurrenceRule,
  now: Date = new Date(),
): Date {
  let next = new Date(due);
  for (let i = 0; i < 400 && next.getTime() <= now.getTime(); i += 1) {
    next = stepLocal(next, rule);
  }
  return next;
}

export function applyRecurrenceAdvances(
  todos: { due_at: string | null; recurrence_rule?: RecurrenceRule | null; checked: boolean }[],
  now: Date = new Date(),
): { todos: typeof todos; changedIndexes: number[] } {
  const changedIndexes: number[] = [];
  const next = todos.map((todo, index) => {
    if (
      !needsRecurrenceAdvance(todo.due_at, todo.recurrence_rule, todo.checked, now)
    ) {
      return todo;
    }
    const rule = todo.recurrence_rule;
    if (!todo.due_at || !rule) return todo;
    changedIndexes.push(index);
    return {
      ...todo,
      due_at: nextRecurringDue(new Date(todo.due_at), rule, now).toISOString(),
    };
  });
  return { todos: next, changedIndexes };
}

export function needsRecurrenceAdvance(
  dueAt: string | null | undefined,
  rule: RecurrenceRule | null | undefined,
  checked: boolean,
  now: Date = new Date(),
): boolean {
  if (checked || !dueAt || !rule) return false;
  return new Date(dueAt).getTime() <= now.getTime();
}
